### PR TITLE
Optimize parallel resource usage and remove DRMAA

### LINKED ISSUE #102: Improve Mutect2 workflow memory usage
**Issue**
We currently have the challenge that the `mutect2` wrapper uses a lot of memory for focal amplifications - rounded up memory usage set to 120GB, which then blocks lots of resources.

**Suggestion**
Snakemake allows to give a callable function as resource instead of a fixed value. This function is given by parameter (maybe others?) the number of retries.  It would be smarter to change this to have the `mutect2` wrapper (which itself is a snakemake call) restart up to 5 times and scale up from 2 GB to 120GB linearly.
Code highlights: 
* `ParallelSomaticVariantCallingBaseWrapper` [#L890](https://github.com/bihealth/snappy-pipeline/blob/7eb21a889da4ec2a54e0d21abec0d994bba9d12f/snappy_wrappers/wrapper_parallel.py#L890)
* `run_snakemake()` [#L236](https://github.com/bihealth/snappy-pipeline/blob/7eb21a889da4ec2a54e0d21abec0d994bba9d12f/snappy_wrappers/wrapper_parallel.py#L236)

**Additional information**
Changes to `mutect2` cannot break `mutect` running. Although we are considering discontinue its use, we still need it for rerunning analysis if necessary.

_Mutect2 Snakemake workflow:_
dag_mutect2

**Required steps**
- [x] Implement unit tests for `mutect2` workflow in `workflows/somatic_variant_calling/__init__.py`.
- [x] Implement unit tests for `ParallelSomaticVariantCallingBaseWrapper` and underlining classes.
- [x] Implement linear memory increase for snakemake call.

### LINKED ISSUE #110: Switch away from DRMAA to using Snakemake Profiles
**Is your feature request related to a problem? Please describe.**
Using cluster configuration in snakemake is deprecated. Profiles are the future.

**Describe the solution you'd like**
We should switch away from using DRMAA towards using Snakemake Profiles. In particular, the [slurm](https://github.com/Snakemake-Profiles/slurm) profile is a target.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by eudesbarbosa**
closes #102, closes #110 

**Changes**
* Removed DRMAA dependency, replaced with Snakemake profile.
* Implemented linear memory increase for Mutect2 variant calling jobs.
* Modified jannovar version: v0.26 -> v0.34.
* Expanded unit test coverage.
* Updated institutional e-mail of authors.

**Tests**
- Passed internal E2E (under development; commit: 96ca804eed1e0f6365eeb2102e89c9362e58afb2)
- 100% agreement using GATK variant caller comparison using `VariantEval`:
```
gatk VariantEval    -R /path/to/reference/GRCh37/hs37d5/hs37d5.fa   \
                              -O NA12878_gatk_hc_output.eval.txt    \
                              --eval /path/to/102-110-mutect2-mem-smk-profiles/bwa.gatk_hc.NA12878-N1-DNA1-WES1.vcf.gz    \
                              --comp /path/to/master/bwa.gatk_hc.NA12878-N1-DNA1-WES1.vcf.gz
```
- Linearly increased memory availability for difficult production case. Most chunks in the parallel Mutect2 call required way less than the minimum memory reserved for each process. We can could decrease the minimum value and increase the number of restarts to deal with the outliers.

mem_usage

### CODE REVIEW

**Comment by eudesbarbosa** [comment]
**Tasks**

- [ ] Move `ResourceUsage` away from Abstract, and adjust all imports.
- [ ] Output slurm log into `<work_dir>/slurm_log/<main_slurm_job_id>/` instead of working directory.
- [ ] Adjust remaining parallel wrappers.

**Comment by eudesbarbosa** [comment]
**Questions**
1. Can someone help me figure out slurm log dir? The slurm log should go into <work_dir>/slurm_log/<main_slurm_job_id>/ instead of working directory. Is this something that can be solved in `wrapper_parallel` or is it on the Snakemake profile side?
2. I removed the `self-test` because its main function was to check if DRMAA was available. If we think we want that back I can right a ticket.
3. While running several cancer workflows at the same time I got several `IndexError` from `slurm-sidecar` - [profile_error.log](https://github.com/bihealth/snappy-pipeline/files/8868769/profile_error.log) . Is there an easy way to deal with  it, like decrease the number of times it can query the queue?

**Review Suggestion**
@All
Documentation: I went for the lazy approach and just removed DRMAA mentions. 

@holtgrewe  and @messersc, please focus on:
- `snappy_pipeline/apps/snappy_snake.py`
- `snappy_pipeline/apps/tpls/pipeline_job.sh`
- `snappy_pipeline/workflows/abstract/__init__.py`
- `snappy_wrappers/wrapper_parallel.py`
- `snappy_wrappers/wrappers/gatk_hc_par`
- `snappy_wrappers/wrappers/mutect2_par`

@mbenary and @ericblanc20 please chose one or two of your most used cancer workflows and give them a check. Skip `cbio_portal`, already covered in #127

**Comment by holtgrewe** [comment]
Quick answer to question 1. I believe that we should not collect the parallel wrapper logs. We should keep the full workflow directory (in the temporary) directory based on the `keep_tmpdir` setting.

**Comment by eudesbarbosa** [comment]
> Quick answer to question 1. I believe that we should not collect the parallel wrapper logs. We should keep the full workflow directory (in the temporary) directory based on the `keep_tmpdir` setting.

Totally fine. Do you know  where is this parameter defined, if in `wrapper_parallel` or is it on the Snakemake profile side? I couldn't find a solution by myself.

**Comment by holtgrewe** [comment]
> Totally fine. Do you know where is this parameter defined, if in `wrapper_parallel` or is it on the Snakemake profile side? I couldn't find a solution by myself.

Oh, I see now. The snakemake profile author says the way to do is using `SBATCH_DEFAULTS`.

https://github.com/Snakemake-Profiles/slurm/issues/90

Hm, having this centrally in the profile might not be the best place. I need to think a bit how to best do this...

**Comment by messersc** [comment]
Re 3: issue is https://github.com/Snakemake-Profiles/slurm/issues/96 and internal discussion at 
https://hpc-talk.cubi.bihealth.org/t/list-index-out-of-range-error-on-snakemake/220/9

not resolved yet.

File: snappy_wrappers/wrapper_parallel.py [hunk 1]
```diff
@@ -112,103 +111,101 @@ def __init__(self, input_files, output_files, resource_usage, script, log_dir=No
...
-    """Return ``datetime.timedelta`` for ``num`` hours"""
-    return datetime.timedelta(hours=num)
+def kib_to_string(num):
+    """
+    :param num: Number of kilobytes.
+    :type num: int
 
+    :return: Returns kilobytes string for memory usage with profile. Example: '200k'.
+    """
+    return f"{int(num)}k"
```
Comment (holtgrewe) [inline snappy_wrappers/wrapper_parallel.py:153]: Should we use `math.ceil` here?

**Review by holtgrewe — APPROVED** [review]
- `snappy_pipeline/apps/snappy_snake.py`: LGTM
- `snappy_pipeline/apps/tpls/pipeline_job.sh`: LGTM
- `snappy_pipeline/workflows/abstract/__init__.py`: LGTM
- `snappy_wrappers/wrapper_parallel.py`: LGTM, added one minor comment
- `snappy_wrappers/wrappers/gatk_hc_par`: LGTM
- `snappy_wrappers/wrappers/mutect2_par`: LGTM, added one request for changes at some point in the future to make generated code better testable

**Review by messersc — APPROVED** [review]
I skimmed the overall PR and closely read the parts you highlighted specifically @eudesbarbosa.

File: snappy_wrappers/wrapper_parallel.py (hunk 1 above)
Comment (eudesbarbosa) [reply snappy_wrappers/wrapper_parallel.py]: Makes sense. Updated.

File: snappy_wrappers/wrapper_parallel.py [hunk 2]
```diff
@@ -112,103 +111,101 @@ def __init__(self, input_files, output_files, resource_usage, script, log_dir=No
...
-    """Converter for Slurm"""
+    :param num: Number of hours.
+    :type num: int
+    :return: Returns ``datetime.timedelta`` as string for ``num`` hours.
+    """
+    output = str(datetime.timedelta(hours=num))
+    if "day" in output:
+        output = output.replace(" days, ", "-").replace(" day, ", "-")
+        tmp_arr = output.split("-")
+        output = tmp_arr[0] + "-" + ":".join([str(i).zfill(2) for i in tmp_arr[1].split(":")])
```
[the remark this answers is not in the document]
Comment (eudesbarbosa) [reply snappy_wrappers/wrapper_parallel.py:202]: Updated.

File: snappy_wrappers/wrappers/mutect2_par/run/parallel_mutect2.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            time=hours(1 * self.get_merge_mult_time()),
+            partition=os.getenv("SNAPPY_PIPELINE_PARTITION"),
+        )
+
+    def construct_preamble(self):
+        """Return a preamble that redefines resource_chunk_{threads,memory} to
+        define functions as "scaling up" with the number of attempts.
+        """
+        return (
+            textwrap.dedent(
```
[the remark this answers is not in the document]
Comment (eudesbarbosa) [reply snappy_wrappers/wrappers/mutect2_par/run/parallel_mutect2.py:72]: Created.

**Pull request merged**